Source/WebGPU/WGSL/GlobalVariableRewriter.cpp

```diff
@@ -0,0 +1,281 @@
+    RewriteGlobalVariables(CallGraph& callGraph, const HashMap<String, PipelineLayout>& pipelineLayouts)
+        : AST::Visitor()
+        , m_callGraph(callGraph)
+    {
+        UNUSED_PARAM(pipelineLayouts);
+    }
+
+    void run();
+
+    void visit(AST::FunctionDecl&) override;
+    void visit(AST::VariableDecl&) override;
+    void visit(AST::IdentifierExpression&) override;
+
+private:
+    template<typename Value>
+    using IndexMap = HashMap<unsigned, Value, WTF::IntHash<unsigned>, WTF::UnsignedWithZeroKeyHashTraits<unsigned>>;
+
+    using IndexSet = HashSet<unsigned, WTF::IntHash<unsigned>, WTF::UnsignedWithZeroKeyHashTraits<unsigned>>;
+
+    struct Global {
+        struct Resource {
+            unsigned m_group;
+            unsigned m_binding;
+        };
+
+        std::optional<Resource> m_resource;
+        AST::VariableDecl* m_declaration;
+    };
+
+    static String argumentBufferParameterName(unsigned group);
+    static String argumentBufferStructName(unsigned group);
+
+    void def(const String&);
+    Global* read(const String&);
+
+    void collectGlobals();
+    void insertStructs();
+    void visitEntryPoint(AST::FunctionDecl&);
+    IndexSet requiredGroups();
+    void insertParameters(AST::FunctionDecl&, const IndexSet& requiredGroups);
+
+    CallGraph& m_callGraph;
+    HashMap<String, Global> m_globals;
+    IndexMap<IndexMap<Global*>> m_groupBindingMap;
+    HashSet<String> m_defs;
+    HashSet<String> m_reads;
+};
+
+void RewriteGlobalVariables::run()
+{
+    collectGlobals();
+    insertStructs();
+    for (auto& entryPoint : m_callGraph.entrypoints())
+        visitEntryPoint(entryPoint.m_function);
+    m_callGraph.ast().globalVars().clear();
+}
+
+void RewriteGlobalVariables::visit(AST::FunctionDecl& functionDecl)
+{
+    // FIXME: visit callee once we have any
+
+    for (auto& parameter : functionDecl.parameters())
+        def(parameter.name());
+
+    // FIXME: detect when we shadow a global that a callee needs
+    AST::Visitor::visit(functionDecl.body());
+}
+
+void RewriteGlobalVariables::visit(AST::VariableDecl& variableDecl)
+{
+    def(variableDecl.name());
+}
+
+template <typename TargetType, typename CurrentType, typename... Arguments>
+void replace(CurrentType& dst, Arguments&&... arguments)
+{
+    static_assert(sizeof(TargetType) <= sizeof(CurrentType));
+    new (&dst) TargetType(dst.span(), std::forward<Arguments>(arguments)...);
+}
+
+void RewriteGlobalVariables::visit(AST::IdentifierExpression& identifier)
+{
+    auto name = identifier.identifier();
+    if (Global* global = read(name)) {
+        if (auto resource = global->m_resource) {
+            replace<AST::StructureAccess>(identifier,
+                makeUniqueRef<AST::IdentifierExpression>(identifier.span(), argumentBufferParameterName(resource->m_group)),
+                WTFMove(name));
+        }
+    }
+}
+
+void RewriteGlobalVariables::collectGlobals()
+{
+    auto& globalVars = m_callGraph.ast().globalVars();
```

Comment:
Maybe this is nit-picking, but it seems counter-intuitive to me that the set of global variables would be a member of the AST. I'd expect the ownership model to be that both the AST and any derived info (like this global variable collection) would be siblings, owned by the same root object.

Reply:
But globals in this context are source elements, so they must be part of the AST, no? Maybe this is a case of meaning different things by `global`? What is meant here is e.g. when we parse:

```
@group(x) @binding(y) var<uniform> u: U;
```

It has to end up in the AST.